qiskit-ibm-transpiler-mcp-server/src/qiskit_ibm_transpiler_mcp_server/qta.py

```diff
@@ -0,0 +1,301 @@
+from typing import Any, Literal
+
+from qiskit_ibm_transpiler_mcp_server.utils import (
+    load_qasm_circuit,
+    get_backend_service,
+    with_sync,
+)
+
+from qiskit.transpiler import PassManager  # type: ignore[import-untyped]
+from qiskit.qasm3 import dumps  # type: ignore[import-untyped]
+from qiskit_ibm_transpiler.ai.routing import AIRouting  # type: ignore[import-untyped]
+from qiskit_ibm_transpiler.ai.synthesis import AICliffordSynthesis  # type: ignore[import-untyped]
+from qiskit_ibm_transpiler.ai.synthesis import AILinearFunctionSynthesis
+from qiskit_ibm_transpiler.ai.synthesis import AIPermutationSynthesis
+from qiskit_ibm_transpiler.ai.synthesis import AIPauliNetworkSynthesis
+
+import logging
+
+# Configure logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+@with_sync
+async def ai_routing(
```

Comment:
All 5 synthesis/routing functions (`ai_routing`, `ai_clifford_synthesis`, `ai_linear_function_synthesis`, `ai_permutation_synthesis`, `ai_pauli_network_synthesis`) follow nearly identical patterns:
- Validate backend_name
- Get backend service
- Check for success/error
- Load QASM circuit
- Check for success/error
- Create PassManager with specific synthesis pass
- Run PassManager
- Dump to QASM
- Return result or error

This is ~50 lines of duplicated code per function (~250 lines total). Recommendation: Extract a helper function:

```
async def _run_synthesis_pass(
    circuit_qasm: str,
    backend_name: str,
    synthesis_pass_class: type,
    pass_kwargs: dict,
    log_message: str
) -> dict[str, Any]:
    # Common implementation here
```